### PR TITLE
Set up a new shopping list

### LINKED ISSUE #3: 3. As a user, I want to set up a new shopping list so I can start tracking purchased items
## Summary 
A shopping list is a set of items associated with a specific three-word token. We need to allow users to start new lists, so they can save the things they need to buy. To start a new list, we will need to do a few things:

1. Generate a new, unique token
2. Save the token to `localStorage`
3. Show the user the list view

Up to this point, the app has rendered the items in the list associated with the token `my test list`. Before we can allow users to make *new* lists, we have to stop showing them this test list.

## Acceptance criteria
- [x] `@the-collab-lab/shopping-list-utils` is added as a dependency to the project.
- [x] The string `my test list` in `App.jsx` is replaced with `null`, so users are not automatically subscribed to any list.

If a user **doesn’t** already have a token:
- [x] A button in the `Home` component allows them to create a new list
- [x] Clicking the button generates a new token and saves it to `localStorage` using the `setListToken` function in `App.jsx`.
- [x] Once the token has been created and saved, the user is redirected to the `List` view.

If a user **does** already have a token:
- [x] They are automatically redirected to the `List` view.

## Notes
- The  `generateToken` function from the `shopping-list-utils` package should be used to generate the token. More information on that package [its documentation](https://www.npmjs.com/package/@the-collab-lab/shopping-list-utils). Importing it will look something like this:
``` js
import { generateToken } from '@the-collab-lab/shopping-list-utils';
```
- The `setListToken` function takes **one** argument: the new list token.
- Eventually the welcome/home page will look like [this wireframe](https://github.com/the-collab-lab/smart-shopping-list/blob/main/_resources/wireframes/welcome-screen.png). It’s good for your whole team to look at the wireframe now, but be sure not to try to implement all of the visible functionality at once. There will be future stories to add the other features such as joining an existing list.
- For this story, it's sufficient just to have a simple "create list" button. Although you are not required to save the token to Firestore, some teams have opted to also save the token to Firestore when generating a token.

### PR DESCRIPTION
**Pull request by Mama-simba**
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._

## Description

Our main goal is enabling the user creating a new list by clicking on the button at the Home component. If the user already has a token, will be redirected straight away to the List view. However, if the user doesn't have any token, this will be generated when clicking on the button.

## Related Issue

Closes #3 

## Acceptance Criteria

- [x] `@the-collab-lab/shopping-list-utils` is added as a dependency to the project.
- [x] The string `my test list` in `App.jsx` is replaced with `null`, so users are not automatically subscribed to any list.

If a user **doesn’t** already have a token:
- [x] A button in the `Home` component allows them to create a new list
- [x] Clicking the button generates a new token and saves it to `localStorage` using the `setListToken` function in `App.jsx`.
- [x] Once the token has been created and saved, the user is redirected to the `List` view.

If a user **does** already have a token:
- [x] They are automatically redirected to the `List` view.

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Testing Steps / QA Criteria

- From your terminal, pull down this branch with git pull origin do-cp-purchased-items and check that branch out with git checkout do-cp-purchased-items
- Then npm start to launch the app. Check on the Chrome console under Application > localStorage to see if you automatically get a token. Otherwise, click on the /home button and test the redirection to /list.

### CODE REVIEW

**Comment by mohanner-coder** [comment]
Alright

**Review by mohanner-coder — COMMENTED** [review]
Ln 19 of .app.jsx, the setItem is not defined.

**Comment by thatgirldorian** [comment]
> Ln 19 of .app.jsx, the setItem is not defined.

setItem is a method, so it doesn't need to be defined.

File: src/App.jsx [hunk 1]
```diff
@@ -1,61 +1,60 @@
...
+		// 		 * Refer to `api/firebase.js`
+		// 		 */
+		// 		const nextData = getItemData(snapshot);
 
-			/** Finally, we update our React state. */
-			setData(nextData);
-		});
+		// 		/** Finally, we update our React state. */
+		// 		setData(nextData);
+		// 	});
```
Comment (faysvas) [inline src/App.jsx:46]: Why is this section commented out here?

File: src/App.jsx [hunk 2]
```diff
@@ -1,61 +1,60 @@
...
-		'my test list',
-		'tcl-shopping-list-token',
-	);
+
+	const [listToken, setListToken] = useStateWithStorage(null);
+
+	function handleClick() {
+		if (!listToken) {
+			const token = generateToken();
+			localStorage.setItem('tcl-shopping-list-token', token);
```
Comment (faysvas) [inline src/App.jsx]: You don't need to store the token again with `localStorage.setItem` because `setListToken` already does that for you :slightly_smiling_face:

File: src/App.jsx [hunk 3]
```diff
@@ -1,61 +1,60 @@
...
-	 * You'll later set it to `null` by default (since new users do not
-	 * have tokens), and use `setListToken` when you allow a user
-	 * to create and join a new list.
-	 */
-	const [listToken, setListToken] = useStateWithStorage(
-		'my test list',
-		'tcl-shopping-list-token',
-	);
+
+	const [listToken, setListToken] = useStateWithStorage(null);
```
Comment (faysvas) [inline src/App.jsx]: For this to work properly, you need to initialize the `useStateWithStorage` hook with  two parameters. The initial value (in your case null) and the `storageKey` (in your case `'tcl-shopping-list-token'`).
Check out the `useStateWithStorage` hook file, there it describes how this hook is meant to be initialized and used. If you see its description, it sets some state and also copies that state over to local storage.

File: src/App.jsx [hunk 4]
```diff
@@ -1,61 +1,60 @@
...
+	function handleClick() {
+		if (!listToken) {
+			const token = generateToken();
+			localStorage.setItem('tcl-shopping-list-token', token);
+			setListToken(token);
+			navigate('/list');
+		} else {
+			navigate('/list');
+		}
+	}
```
Comment (faysvas) [inline src/App.jsx]: This `handleClick` function should only be called when creating a new list so I don't think it's necessary to have an else that redirects to list.

File: src/App.jsx [hunk 5]
```diff
@@ -1,61 +1,60 @@
...
-			 */
-			const nextData = getItemData(snapshot);
+		// 	return streamListItems(listToken, (snapshot) => {
+		// 		/**
+		// 		 * Read the documents in the snapshot and do some work
+		// 		 * on them, so we can save them in our React state.
+		// 		 *
+		// 		 * Refer to `api/firebase.js`
+		// 		 */
+		// 		const nextData = getItemData(snapshot);
```
Comment (faysvas) [inline src/App.jsx:43]: Why is this section commented out instead of deleted?

File: src/views/Home.jsx [hunk 6]
```diff
@@ -1,11 +1,12 @@
 import './Home.css';
 
-export function Home() {
+export function Home({ handleClick }) {
 	return (
 		<div className="Home">
 			<p>
 				Hello from the home (<code>/</code>) page!
 			</p>
+			<button onClick={handleClick}>Create list</button>
```
Comment (faysvas) [inline src/views/Home.jsx]: This button should be hidden if a `listToken` already exists because that means we already have a list and don't need to create one:slightly_smiling_face:

File: vite.config.js [hunk 7]
```diff
@@ -40,7 +40,10 @@ export default defineConfig({
 	build: {
 		outDir: './build',
 		target: 'esnext',
-		commonjsOptions: { include: [] },
+		commonjsOptions: {
+			include: [],
+			esmExternals: true,
```
Comment (faysvas) [inline vite.config.js]: Why did you need to add this extra field (`esmExternals`) to the vite configuration?

**Review by thepatrickniyo — CHANGES_REQUESTED** [review]
Please @Mama-simba try resolving all the conversations with @faysvas . 
Address the changes requested. Then we will be good to go

File: src/App.jsx (hunk 1 above)
Comment (thatgirldorian) [reply src/App.jsx:46]: This code is commented out because it's part of the boilerplate code that comes with the project. I commented it out in case it relates to another future task. It has now been added back to the code.

File: src/App.jsx (hunk 2 above)
Comment (thatgirldorian) [reply src/App.jsx]: Edited the code to reflect this. :)

File: src/App.jsx (hunk 3 above)
Comment (thatgirldorian) [reply src/App.jsx]: We've updated this as well

File: src/App.jsx (hunk 4 above)
Comment (thatgirldorian) [reply src/App.jsx]: Fixed this!

File: src/App.jsx (hunk 5 above)
Comment (thatgirldorian) [reply src/App.jsx:43]: This is part of the boilerplate code that comes with the project. I didn't want to remove it in case of a future project.

File: vite.config.js (hunk 7 above)
Comment (Mama-simba) [reply vite.config.js]: Not sure about this but I have just removed it

File: src/views/Home.jsx (hunk 6 above)
Comment (Mama-simba) [reply src/views/Home.jsx]: if the user has a token, gets redirected to the List view, so the user won't see the button. Isn't it correct that way?

Comment (thatgirldorian) [reply src/views/Home.jsx]: Just fixed this!

Quick question! Do users need to be redirected to the List view from wherever they are in the app or does this behavior only pertains to the Home view?

**Comment by thepatrickniyo** [comment]
Keep up the great work @Mama-simba and @thatgirldorian but then let's now first remove the conflicts in this branch by pull the changes on the main branch.
This was introduced because of @Terieyenike 's work was merged first. 
Therefore we can merge the work without any version collision.

**Review by thepatrickniyo — APPROVED** [review]

File: src/views/Home.jsx (hunk 6 above)
Comment (faysvas) [reply src/views/Home.jsx]: I'd say it makes sense to only redirect them from the home view, because if they already have a list there is nothing for them to do there. It doesn't make much sense to redirect them from "Add Item" view because they might want to add items.

Comment (Mama-simba) [reply src/views/Home.jsx]: @thatgirldorian has already fixed it :)
